File: src/scoreboard-client/lib/api/leaderboard.ts
import { getJson } from "../http";
import type { LeaderboardResponse } from "../dtos/api/leaderboard";
import type { ScoreScope } from "../score_scope";

export function leaderboard(
  model: string | null,
  view: string,
  scope: ScoreScope,
): Promise<LeaderboardResponse> {
  const params = new URLSearchParams({ view, scope });
  if (model) params.set("model", model);
  return getJson<LeaderboardResponse>(`/api/leaderboard?${params.toString()}`);
}

File: src/scoreboard-server/scoreboard_server/routes/api/leaderboard.py
from __future__ import annotations

from typing import Literal

from fastapi import FastAPI, Query

from scoreboard_server.cores.normalize import DEFAULT_TABLE_VIEW
from scoreboard_server.db.repository import ScoreboardStore
from scoreboard_server.dtos.api.leaderboard import LeaderboardResponse
from scoreboard_server.services.api.leaderboard import leaderboard_response


def register(app: FastAPI, store: ScoreboardStore) -> None:
    @app.get("/api/leaderboard")
    async def leaderboard(
        model: str | None = Query(default=None),
        view: str = Query(default=DEFAULT_TABLE_VIEW),
        scope: Literal["official", "non_official"] = "official",
    ) -> LeaderboardResponse:
        return await leaderboard_response(
            store,
            model=model,
            view=view,
            scope=scope,
        )
